src/frontend/src/binder/set_expr.rs

```diff
@@ -36,6 +36,11 @@ pub enum BoundSetExpr {
         left: Box<BoundSetExpr>,
         right: Box<BoundSetExpr>,
     },
+    /// UNION in recursive CTE definition
+    RecursiveUnion {
+        base: Box<BoundSetExpr>,
+        recursive: Box<BoundSetExpr>,
+    },
```

Comment:
I feel this is not necessary as well. Consider recursive functions in common programming languages, compiler won't have any special representations of the expressions in this function just because it is recursive, right? 🤔

Reply:
But it's a special operator.

RecursiveUnion has extremely different schema/stream-key derivation logical, and have unique executor implementations in all databases.

Reply:
https://github.com/postgres/postgres/blob/master/src/backend/executor/nodeRecursiveunion.c

In postgres

Reply:
https://github.com/duckdb/duckdb/blob/main/src/execution/operator/set/physical_recursive_cte.cpp

In duckdb (UNION is executed in the operator)

Reply:
You will find that in RCTE definition, UNION is special enough. The top level operator of RCTE must be a `UNION` or `UNION ALL`.